client/statement-store/src/lib.rs

```diff
@@ -0,0 +1,1263 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+// Constraint management.
+//
+// Each time a new statement is inserted into the store, it is first validated with the runtime
+// Validation function computes `global_priority`, 'max_count' and `max_size` for a statement.
+// The following constraints are then checked:
+// * For a given account id, there may be at most `max_count` statements with `max_size` total data
+//   size. To satisfy this, statements for this account ID are removed from the store starting with
+//   the lowest priority until a constraint is satisfied.
+// * There may not be more than `MAX_TOTAL_STATEMENTS` total statements with `MAX_TOTAL_SIZE` size.
+//   To satisfy this, statements are removed from the store starting with the lowest
+//   `global_priority` until a constraint is satisfied.
+//
+// When a new statement is inserted that would not satisfy constraints in the first place, no
+// statements are deleted and `Ignored` result is returned.
+// The order in which statements with the same priority are deleted is unspecified.
+//
+// Statement expiration.
+//
+// Each time a statement is removed from the store (Either evicted by higher priority statement or
+// explicitly with the `remove` function) the statement is marked as expired. Expired statements
+// can't be added to the store for `PURGE_AFTER` seconds. This is to prevent old statements from
+// being propagated on the network.
+
+//! Disk-backed statement store.
+
+#![warn(missing_docs)]
+#![warn(unused_extern_crates)]
+
+mod metrics;
+
+pub use sp_statement_store::{Error, StatementStore};
+
+use metrics::MetricsLink as PrometheusMetrics;
+use parking_lot::RwLock;
+use prometheus_endpoint::Registry as PrometheusRegistry;
+use sp_api::ProvideRuntimeApi;
+use sp_blockchain::HeaderBackend;
+use sp_core::{hexdisplay::HexDisplay, Decode, Encode};
+use sp_runtime::traits::Block as BlockT;
+use sp_statement_store::{
+	runtime_api::{InvalidStatement, StatementSource, ValidStatement, ValidateStatement},
+	AccountId, BlockHash, Channel, DecryptionKey, Hash, NetworkPriority, Proof, Result, Statement,
+	SubmitResult, Topic,
+};
+use std::{
+	collections::{BTreeMap, HashMap, HashSet},
+	sync::Arc,
+};
+
+const KEY_VERSION: &[u8] = b"version".as_slice();
+const CURRENT_VERSION: u32 = 1;
+
+const LOG_TARGET: &str = "statement-store";
+
+const PURGE_AFTER: u64 = 2 * 24 * 60 * 60; //48h
+const MAX_TOTAL_STATEMENTS: usize = 8192;
+const MAX_TOTAL_SIZE: usize = 64 * 1024 * 1024;
+
+/// Suggested maintenance period. A good value to call `Store::maintain` with.
+#[allow(dead_code)]
```

Comment:
pub should not be dead code?

Reply:
Why isn't the maintenance loop just implemented in this crate?

Reply:
Because it does not have the responsibility to manage workers. I did not want to introduce yet another point where we spawn more workers.

Reply:
Moved to `sc-statement-store`